master/internal/stream/system.go

```diff
@@ -191,9 +207,104 @@ func writeAll(socketLike WebsocketLike, msgs []interface{}) error {
 	return nil
 }
 
+// processStream processes as startup message, then streams live updates until either:
+// - another startup message arrives, in which case it returns it or
+// - the streamer is closed gracefully, in which case it returns nil, or
+// - an error occurs.
+func (ps *PublisherSet) processStream(
+	ctx context.Context,
+	streamer *stream.Streamer,
+	user model.User,
+	startupMsg StartupMsg,
+	startups *[]StartupMsg,
+	prepare func(message stream.PreparableMessage) interface{},
+	socket WebsocketLike,
+) (
+	nextStartup *StartupMsg, err error,
+) {
+	// create new subscription set
+	ss, err := NewSubscriptionSet(ctx, streamer, ps, user, startupMsg.Subscribe)
+	if err != nil {
+		return nil, errors.Wrap(err, "creating subscription set")
+	}
+	defer ss.UnregisterAll()
+
+	// startup subscription set
+	msgs := []interface{}{}
+	offlineMsgs, err := ss.Startup(ctx, user, startupMsg)
+	if err != nil {
+		return nil, errors.Wrapf(err, "gathering startup messages")
+	}
+	msgs = append(msgs, offlineMsgs...)
+
+	// always include a sync message
+	syncMsg := SyncMsg{SyncID: startupMsg.SyncID}
+	msgs = append(msgs, prepare(syncMsg))
+
+	// write offline msgs to the websocket
+	err = writeAll(socket, msgs)
+	if err != nil {
+		// don't log broken pipe errors
+		if !websocket.IsCloseError(err, websocket.CloseAbnormalClosure) {
+			log.Debugf("unable to handle startup message: %s", err)
+		}
+		return nil, errors.Wrapf(err, "error handling startup message")
+	}
+
+	// startup done, begin streaming supported online events:
+	//	- insertions
+	//	- updates
+	//	- deletions
+	//	- fallin
+	//	- fallout
+	// note: online appearances and disappearances are not supported; we'll detect those
+	// situtations and break the connection to the streaming clients
+
+	// waitForSomething Returns a tuple of (first-startup-msg, msgs, closed)
+	waitForSomething := func() (*StartupMsg, []interface{}, bool) {
+		streamer.Cond.L.Lock()
+		defer streamer.Cond.L.Unlock()
+
+		for len(*startups) == 0 && len(streamer.Msgs) == 0 && !streamer.Closed {
+			streamer.Cond.Wait()
+		}
+		// steal outputs
+		var startup *StartupMsg
+		if len(*startups) > 0 {
+			startup = &(*startups)[0]
+			*startups = (*startups)[1:]
+		}
+		msgs := streamer.Msgs
+		streamer.Msgs = nil
+		return startup, msgs, streamer.Closed
+	}
+	for {
+		startup, msgs, closed := waitForSomething()
+
+		// is the streamer closed?
+		if closed {
+			return nil, nil
+		}
+
+		// soft reset?
+		if startup != nil {
+			return startup, nil
+		}
+
+		// otherwise write all the messages we just got
+		err = writeAll(socket, msgs)
+		if err != nil {
+			if !websocket.IsCloseError(err, websocket.CloseAbnormalClosure) {
+				log.Debugf("unable to handle startup message: %s", err)
```

Comment:
You have a good point! Since this impacts functionality, there is an argument to be made for this to be at the error level instead of the debug level.

Do we have expectations around these different log levels beyond:
- Debug: detailed debugging information 
- Error: errors that are relevant to users and helpful for identifying why failures occurred?